edgelet/edgelet-kube/src/module/authentication.rs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::cell::RefCell;
+use std::sync::{Arc, Mutex};
+
+use futures::future::{Either, IntoFuture};
+use futures::{future, Future, Stream};
+use hyper::service::Service;
+use hyper::{Body, Request};
+use log::Level;
+use typed_headers::{Authorization, HeaderMapExt};
+
+use edgelet_core::AuthId;
+use edgelet_utils::log_failure;
+use kube_client::{Client as KubeClient, Error as KubeClientError, TokenSource};
+
+use crate::constants::EDGE_ORIGINAL_MODULEID;
+use crate::error::Error;
+use crate::{KubeModuleRuntime, KubeRuntimeData};
+
+pub fn authenticate<T, S>(
+    runtime: &KubeModuleRuntime<T, S>,
+    req: &Request<Body>,
+) -> impl Future<Item = AuthId, Error = Error>
+where
+    T: TokenSource + 'static,
+    S: Service + Send + 'static,
+    S::ReqBody: From<Vec<u8>>,
+    S::ResBody: Stream,
+    Body: From<S::ResBody>,
+    S::Error: Into<KubeClientError>,
+    S::Future: Send,
+{
+    req.headers()
+        .typed_get::<Authorization>()
+        .map(|auth| {
+            auth.and_then(|auth| {
+                auth.as_bearer().map(|token| {
+                    let client_copy = runtime.client();
+                    let namespace = runtime.namespace().to_owned();
+                    let fut = runtime
+                        .client()
+                        .lock()
+                        .expect("Unexpected lock error")
+                        .borrow_mut()
+                        .token_review(runtime.namespace(), token.as_str())
+                        .map_err(|err| {
+                            log_failure(Level::Warn, &err);
+                            Error::from(err)
+                        })
+                        .and_then(move |token_review| {
+                            token_review
+                                .status
+                                .as_ref()
+                                .filter(|status| status.authenticated.filter(|x| *x).is_some())
+                                .and_then(|status| {
+                                    status.user.as_ref().and_then(|user| user.username.clone())
+                                })
+                                .map_or(Either::A(future::ok(AuthId::None)), |name| {
+                                    Either::B(get_module_original_name(
```

Comment:
Should we be doing any normalization of the module name? Does uppercase vs lowercase matter?

Reply:
The idea was to preserve the original module name as is. It is stored as a service account annotation value. I think no normalization/sanitizing is needed.

Reply:
We ran into this problem a lot during the preview.  The "name" for the service account needs to fit the [name criteria](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names), which doesn't match what edge allows for module names.  For all Kubernetes primitives, we've been putting the original module name as an annotation.  I think what Denis is doing is fine.